text/separate-compilation.md

```diff
@@ -0,0 +1,229 @@
+Export __lib. (* make the declarations from lib available from [Import]ing lib (rather that lib.lib *)
+```
+
+In this example translation, we use the `\_\_` prefix for generated identifiers
+used as a translation device; these identifiers should be hidden from clients.
+
+This translation turns the `.vi` interface into a module type, the `.v`
+implementation into a module, and uses opaque module ascription to hide the
+module contents.
+The only visible side effects are those that appear in the interface: here,
+only `Global Open Scope Z_scope`. As we only `Require` stdpp.prelude in the
+module body, we intend this to be **hidden** from clients that perform `Require
+lib`, even if this might not be guaranteed by `lib_composed.v`.
+
+## Only a `.vi` File
+
+When only an interface file exists, there is (potentially) no underlying
+implementation, but the existance of the interface should still provide definite
+references to an implementation. Casting this in the feature set of the current
+Coq ecosystem, this essentially translates to a `Declare Module`
+[^full-functorization]. Concretely,
+
+```coq
+(* lib.vi *)
+Global Open Scope Z_scope.
+Parameter answer : nat.
+Axiom answer_is_42 : answer = 42.
+```
+
+would be "translated" to:
+
+```coq
+(* lib.v *)
+Module Type __LIB.
+  Global Open Scope Z_scope.
+  Parameter answer : nat.
+  Axiom answer_is_42 : answer = 42.
+End __LIB.
+
+Declare Module __lib : __LIB.
+Export __lib.
+```
+
+## Only a `.v` File
+
+If only a `.v` file is available, Coq's semantics should be the same as today:
+the entire implementation shall be exposed in full builds, and it shall be
+hidden in `vos` builds.
+
+## Universes
+
+As some readers will anticipate, universe checks do not admit fully separate
+compilation. Module bodies might add universe constraints that are note explicit
+(or even visible) from its interface. This is already an issue with `.vos`
+builds today, and is a problem inherent to Coq's global universe graph and
+(generally) implicit treatment of universes in the surface syntax, so any
+solution to this problem could be shared.
+
+We consider `vos` builds a special case of this proposal, where interfaces are
+inferred as the strictest possible ones for the given implementation; `.vi`
+files enable hiding more implementation details. In both cases, the interface
+omits universe constraints that are derived from hidden bodies (for `vos`
+builds, just `Qed` bodies). While some universe checks are performed anyway,
+omitted constraints might make the universe graph unsatisfiable.
+
+To remedy this problem, we propose an additional "global" check. By analogy with
+separate compilation in other languages, we call this "link-time" universe
+checking. Such a link-time would be effectively the same as `Require`ing all of
+the modules in a library preferring their *implemenation* (`.v` file) rather
+than their interface (`.vi` file). If the combined set of constraints is
+satisfiable, then there is no problem.
```

Comment:
The problem with universes is not so much satisfiability than the fact you can get wildly different results depending on the link order, because this affects unification and the like. It is somewhat easy to craft a `Definition foo : bool.` whose value in separate compilation mode depends on the scheduling order.

Reply:
> because this affects unification and the like.

> It is somewhat easy to craft a Definition foo : bool. whose value in separate compilation mode depends on the scheduling order.

Link-time checks happen on concrete proof terms, so what's the unification in question?

Reply:
The issue I mentioned happens for vos files, but a similar problem arises with separate compilation. Namely, there are cases where replacing the interface with the implementation will result in later files being compiled differently, breaking the abstraction boundary. For a simplified and contrived example, you can consider
```coq
Universe i j.

Lemma dummy : True.
Proof.
constr_eq Type@{i} Type@{j}.
constructor.
Qed.

Definition b : bool.
Proof.
(let _ := constr:(let _ : Type@{i} := Type@{j} in tt) in exact true) || exact false.
Defined.
```
where the value of `b` depends whether the proof script of `dummy` has been evaluated or not at the time `b` is computed. In separate compilation mode, it means that an interface where `dummy` is an axiom won't produce the same proof as one where the proof is explicitly given, despite being clearly compatible as per the module rules. This is not specific to Qed proofs, one can also trigger this kind of behaviours with explicit constraints hidden under module interfaces.

Reply:
But in the separate compilation model we don't expect to be able to remove the interface and still compile, so it's not a problem IMO

Reply:
Yes, “removing the interface breaks clients” is an intended feature. But the only existing way to fully compile this code in non-vos mode would leak the extra constraint to clients, despite opaque ascriptions; this was brought up in December and it’s concerning. I’ve seen no evidence this is a real problem (compared to the periodic universe inconsistencies we do see), but in principle it seems fixable (very invasively!) if you’re that serious about separate compilation: always defer checking of these constraints to the link-time check. But that’s independent of this proposal, and relevant to `vos` builds in general.

> despite being clearly compatible as per the module rules

I agree that the subsumption rule would suggest ”dropping interfaces” should work, but subsumption on source-language modules doesn’t actually work (in Coq, OCaml and many other languages). Even ignoring hints etc., `Import` + width subtyping is enough to break subsumption: if you `Import foo` and `foo` adds a member `bar`, then `foo` changes to a more specific type, yet `bar` can shadow existing `bar` and break clients.